File: 백준/Gold/21608. 상어 초등학교/상어 초등학교.py
n = int(input())
seats = [[0]*(n+1) for _ in range(n+1)]
like_list = [0]*(n*n+1)
order = []

# 상좌우하순
dr = [-1, 0, 0, 1]
dc = [0, 1, -1, 0]

# 좋아하는 사람 리스트와 순서 저장
for _ in range(n*n):
  num = list(map(int, input().split()))
  like_list[num[0]] = num[1:]
  order.append(num[0])

# 학생 자리 배치
for now in order:
  temp = [] # 현재 학생을 앉힐 수 있는 자리 후보군
  for i in range(1, n+1):
    for j in range(1, n+1):
      if seats[i][j] == 0:   # 빈칸일 경우
        like_cnt = 0  # 좋아하는 학생 수
        empty_cnt = 0  # 인접한 빈칸 수

        for k in range(4):  # 4방향 탐색
          nr = i + dr[k]
          nc = j + dc[k]
          
          if 1<=nr<n+1 and 1<=nc<n+1:   # 좌석 범위일 때
            if seats[nr][nc] in like_list[now]:   # 인접한 칸에 좋아하는 사람이 있다면
              like_cnt += 1
            if seats[nr][nc] == 0: # 인접한 칸이 비었다면
              empty_cnt += 1
        
        # 후보군에 추가 [인접한 좋아하는 학생수, 인접한 빈칸 수, 현재 행, 현재 열]
        temp.append([like_cnt, empty_cnt, i, j])

  # 리스트를 정렬 (like_cnt와 empty_cnt는 내림차순, 행과 열은 오름차순)
  temp.sort(key = lambda x:(-x[0], -x[1], x[2], x[3]))
  
  # 최적의 조건에 학생 배치 (행과 열)
  seats[temp[0][2]][temp[0][3]] = now

# 점수 계산
res = 0
for i in range(1, n+1):
  for j in range(1, n+1):
    now = seats[i][j] 
    cnt = 0
    
    for k in range(4):
      nr = i + dr[k]
      nc = j + dc[k]
      # 범위 안이고, 인접한 학생이 좋아하는 학생이라면
      if 1<=nr<n+1 and 1<=nc<n+1 and seats[nr][nc] in like_list[now]:  
          cnt += 1
      
    if cnt != 0:
      res += 10**(cnt-1)

print(res)

File: 백준/Gold/21608. 상어 초등학교/상어 초등학교.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
	static int n;
	static int[] order;
	static int[][] like, seats;
	static int[] dx = {0, 0, 1, -1};
	static int[] dy = {1, -1, 0, 0};
	
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;
        
    	n = Integer.parseInt(br.readLine());
    	order = new int[n*n];
    	like = new int[n*n+1][4];
    	seats = new int[n*n][n*n];
        
    	for (int i=0; i<n*n; i++) {
    		String[] str = br.readLine().split(" ");
    		int row = Integer.parseInt(str[0]);
    		order[i] = row;
    		int col = 1;
    		for (int j=0; j<4; j++) {
    			like[row][j] = Integer.parseInt(str[col++]);
    		}
    	}
    	
    	// 다중정렬을 이용한 자리 앉히기
    	sitting();
    	
    	// 점수 계산
    	int sum = 0;
    	for (int i=0; i<n; i++) {
    		for (int j=0; j<n; j++) {
    			int now = seats[i][j];
    			int likeCnt = 0;
    			
    			for (int d=0; d<4; d++) {
    				int nx = i+dx[d];
    				int ny = j+dy[d];
    				
    				if (0<=nx && nx<n && 0<=ny && ny<n) {
						for (int l : like[now]) {
							if (seats[nx][ny] == l) {
								likeCnt++;
							}
						}
    				}
    			}
    			if (likeCnt != 0) {
    			    sum += Math.pow(10 ,likeCnt-1);
    			}
    		}
    	}
    	
    	System.out.print(sum);
    }
    
    private static void sitting() {
    	for (int now : order) {
    		List<int[]> state = new ArrayList<>();
    		
    		for (int i=0; i<n; i++) {
    			for (int j=0; j<n; j++) {
    				if (seats[i][j] == 0) {
    					int likeCnt = 0;
    					int emptyCnt = 0;
    					
    					for (int d=0; d<4; d++) {
    						int nx = i+dx[d];
    	    				int ny = j+dy[d];
    						
    						if (0<=nx && nx<n && 0<=ny && ny<n) {
    							for (int l : like[now]) {
    								if (seats[nx][ny] == l) {
    									likeCnt++;
    								}
    							}
    							if(seats[nx][ny] == 0) {
    								emptyCnt++;
    							}
    						}
    					}
    					state.add(new int[]{likeCnt, emptyCnt, i, j});
    				}
    			}
    		}
    		
    		state.sort((o1, o2) -> {
    		    if (o1[0] != o2[0]) return o2[0] - o1[0]; // 좋아하는 학생수 내림차순
    		    else if (o1[1] != o2[1]) return o2[1] - o1[1]; // 인접한 칸수 내림차순
    		    else if (o1[2] != o2[2]) return o1[2] - o2[2]; // 행 오름차순
    		    else return o1[3] - o2[3]; // 열 오름차순
    		});

    		
    		seats[state.get(0)[2]][state.get(0)[3]] = now;
    	}
    }
}